### PR TITLE
Improve Google Calendar APIs using OAS

### PR DESCRIPTION
**Pull request by Nuvindu**
# Description

- Improve connector APIs using OAS
- Fixed by: [#5712](https://github.com/ballerina-platform/ballerina-standard-library/issues/5712)

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

**Review by TharmiganK — APPROVED** [review]

**Comment by ThisaruGuruge** [comment]
Shall we squash the commits?

File: ballerina/client.bal [hunk 1]
```diff
@@ -0,0 +1,743 @@
...
+    # + 'key - API key. Your API key identifies your project and provides you with API access, quota, and reports. Required unless you provide an OAuth 2.0 token.
+    # + oauth_token - OAuth 2.0 token for the current user.
+    # + prettyPrint - Returns response with indentations and line breaks.
+    # + quotaUser - An opaque string that represents a user for quota purposes. Must not exceed 40 characters.
+    # + payload - Data required to create the calendar.
+    # + return - Successful response 
+    resource isolated function post calendars(Calendar payload, "json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = ()) returns Calendar|error {
+        string resourcePath = string `/calendars`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
```
Comment (dilanSachi) [inline ballerina/client.bal:68]:
```suggestion
        string resourcePath = string `/calendars${check getPathForQueryParam(queryParam)}`;
```

File: ballerina/client.bal [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+    # + payload - Data required to update the calendar.
+    # + return - Successful response 
+    resource isolated function put calendars/[string calendarId](Calendar payload, "json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = ()) returns Calendar|error {
+        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
+        http:Request request = new;
+        json jsonBody = payload.toJson();
+        request.setPayload(jsonBody, "application/json");
+        Calendar response = check self.clientEp->put(resourcePath, request);
```
Comment (dilanSachi) [inline ballerina/client.bal:112]:
```suggestion
        Calendar response = check self.clientEp->put(resourcePath, request);
```
```suggestion
        return self.clientEp->put(resourcePath, request);
```

File: ballerina/client.bal [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+    # + oauth_token - OAuth 2.0 token for the current user.
+    # + prettyPrint - Returns response with indentations and line breaks.
+    # + quotaUser - An opaque string that represents a user for quota purposes. Must not exceed 40 characters.
+    # + calendarId - Calendar identifier. To retrieve calendar IDs call the calendarList.list method. If you want to access the primary calendar of the currently logged in user, use the "primary" keyword.
+    # + return - Successful response 
+    resource isolated function delete calendars/[string calendarId]("json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = ()) returns error? {
+        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
+        return check self.clientEp-> delete(resourcePath);
```
Comment (dilanSachi) [inline ballerina/client.bal:130]:
```suggestion
        return self.clientEp-> delete(resourcePath);
```

File: ballerina/client.bal [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+    # + calendarId - Calendar identifier. To retrieve calendar IDs call the calendarList.list method. If you want to access the primary calendar of the currently logged in user, use the "primary" keyword.
+    # + payload - Data required to update the calendar.
+    # + return - Successful response 
+    resource isolated function patch calendars/[string calendarId](Calendar payload, "json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = ()) returns Calendar|error {
+        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
+        http:Request request = new;
+        json jsonBody = payload.toJson();
+        request.setPayload(jsonBody, "application/json");
```
Comment (dilanSachi) [inline ballerina/client.bal:150]: let's create a constant for `application/json` as its used multiple times

File: ballerina/client.bal [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+    # + maxResults - Maximum number of entries returned on one result page. By default the value is 100 entries. The page size can never be larger than 250 entries. Optional.
+    # + pageToken - Token specifying which result page to return. Optional.
+    # + showDeleted - Whether to include deleted ACLs in the result. Deleted ACLs are represented by role equal to "none". Deleted ACLs will always be included if syncToken is provided. Optional. The default is False.
+    # + syncToken - Token obtained from the nextSyncToken field returned on the last page of results from the previous list request. It makes the result of this list request contain only entries that have changed since then. All entries deleted since the previous list request will always be in the result set and it is not allowed to set showDeleted to False.
+    # If the syncToken expires, the server will respond with a 410 GONE response code and the client should clear its storage and perform a full synchronization without any syncToken.
+    # Learn more about incremental synchronization.
+    # Optional. The default is to return all entries.
+    # + return - Successful response
```
Comment (dilanSachi) [inline ballerina/client.bal:171]:
```suggestion
    # + return - Successful response 
```
Shall we improve the return statements

File: ballerina/client.bal [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+    # + calendarId - Calendar identifier. To retrieve calendar IDs call the calendarList.list method. If you want to access the primary calendar of the currently logged in user, use the "primary" keyword.
+    # + sendNotifications - Whether to send notifications about the calendar sharing change. Optional. The default is True.
+    # + payload - Data required to create access permissions for the calendar.
+    # + return - Successful response 
+    resource isolated function post calendars/[string calendarId]/acl(AclRule payload, "json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = (), boolean? sendNotifications = ()) returns AclRule|error {
+        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}/acl`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser, "sendNotifications": sendNotifications};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
```
Comment (dilanSachi) [inline ballerina/client.bal:195]:
```suggestion
        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}/acl${check getPathForQueryParam(queryParam)}`;
```
We can do like this in all the places right?

File: ballerina/client.bal [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+    # + prettyPrint - Returns response with indentations and line breaks.
+    # + quotaUser - An opaque string that represents a user for quota purposes. Must not exceed 40 characters.
+    # + calendarId - Calendar identifier. To retrieve calendar IDs call the calendarList.list method. If you want to access the primary calendar of the currently logged in user, use the "primary" keyword.
+    # + return - Successful response 
+    resource isolated function post calendars/[string calendarId]/clear("json"? alt = (), string? fields = (), string? 'key = (), string? oauth_token = (), boolean? prettyPrint = (), string? quotaUser = ()) returns error? {
+        string resourcePath = string `/calendars/${getEncodedUri(calendarId)}/clear`;
+        map<anydata> queryParam = {"alt": alt, "fields": fields, "key": 'key, "oauth_token": oauth_token, "prettyPrint": prettyPrint, "quotaUser": quotaUser};
+        resourcePath = resourcePath + check getPathForQueryParam(queryParam);
+        http:Request request = new;
+        return check self.clientEp-> post(resourcePath, request);
```
Comment (dilanSachi) [inline ballerina/client.bal:303]:
```suggestion
        return self.clientEp->post(resourcePath, request);
```

File: ballerina/tests/test_client.bal [hunk 8]
```diff
@@ -0,0 +1,940 @@
...
+        summary: summary
+    };
+    Calendar unionResult = check client1->/calendars.post(cal);
+    test:assertEquals(unionResult.summary, summary);
+
+    error? res = check client1->/calendars/[<string>unionResult.id].delete();
+    test:assertEquals(res, ());
+}
+
+// Define the test configuration
```
Comment (dilanSachi) [inline ballerina/tests/test_client.bal:48]: is this comment related?

File: ballerina/types.bal [hunk 9]
```diff
@@ -0,0 +1,814 @@
...
+    # Type of the resource ("calendar#freeBusy").
+    string kind?;
+    # The end of the interval.
+    string timeMax?;
+    # The start of the interval.
+    string timeMin?;
+};
+
+# Extended properties of the event.
+public type Event_extendedProperties record {
```
Comment (dilanSachi) [inline ballerina/types.bal:100]: some records have extra newline after the record description. i guess we dont need a newline right?

File: examples/build.gradle [hunk 10]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dilanSachi) [inline examples/build.gradle:2]:
```suggestion
 * Copyright (c) 2023, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.

```

File: gradle.properties [hunk 11]
```diff
@@ -0,0 +1,12 @@
+org.gradle.caching=true
+group=io.ballerina.stdlib
+version=0.1.0-SNAPSHOT
```
Comment (dilanSachi) [inline gradle.properties:3]: version?

Comment (Nuvindu) [reply gradle.properties]: Fixed in [0df993b](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/0df993b9d0ae1de867b6aa56ae639eecc72b24b8)

File: ballerina/client.bal (hunk 5 above)
Comment (Nuvindu) [reply ballerina/client.bal]: Fixed in [ed81f7c](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/ed81f7cb7f43ae127ccbee9ef1cade8db0cb7658)

File: examples/work-schedule/main.bal [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+configurable string refreshToken = os:getEnv("REFRESH_TOKEN");
+configurable string refreshUrl = os:getEnv("REFRESH_URL");
+
+public function main() returns error? {
+    calendar:Client calendarClient = check new ({
+        auth: {
+            clientId: clientId,
+            clientSecret: clientSecret,
+            refreshToken: refreshToken,
+            refreshUrl: refreshUrl
```
Comment (ThisaruGuruge) [inline examples/work-schedule/main.bal:32]: Check similar places.

```suggestion
            clientId,
            clientSecret,
            refreshToken,
            refreshUrl
```

File: gradle.properties [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+version=4.0.0-SNAPSHOT
+
+checkstylePluginVersion=10.12.0
+spotbugsPluginVersion=5.0.14
+shadowJarPluginVersion=8.1.1
+downloadPluginVersion=5.4.0
+releasePluginVersion=2.8.0
+testngVersion=7.6.1
+eclipseLsp4jVersion=0.12.0
+ballerinaGradlePluginVersion=2.1.5
```
Comment (ThisaruGuruge) [inline gradle.properties:12]: Let's use the latest version.

File: ballerina/tests/test_client.bal [hunk 14]
```diff
@@ -0,0 +1,693 @@
...
+    Calendar cal = {
+        summary: summary
+    };
+    Calendar createdCal = check client1->/calendars.post(cal);
+    string newSummary = "Updated Test Calendar";
+    createdCal.summary = newSummary;
+    Calendar updatedCal = check client1->/calendars/[<string>createdCal.id].put(createdCal);
+    test:assertEquals(updatedCal.summary, newSummary);
+    Error? res = check client1->/calendars/[<string>createdCal.id].delete();
+    test:assertEquals(res, ());
```
Comment (ayeshLK) [inline ballerina/tests/test_client.bal:75]:
```suggestion
    check client1->/calendars/[<string>createdCal.id].delete();
```

File: ballerina/tests/test_client.bal [hunk 15]
```diff
@@ -0,0 +1,693 @@
...
+            value: "testuser@gmail.com"
+        }
+    };
+    AclRule createdAclRule = check client1->/calendars/[<string>createdCal.id]/acl.post(aclRule);
+    Acl acl = check client1->/calendars/[<string>createdCal.id]/acl.get();
+    boolean aclRuleFound = false;
+    AclRule[]? aclItems = acl.items;
+    if aclItems != () {
+        foreach AclRule acl_item in aclItems {
+            if (acl_item.id == createdAclRule.id) {
```
Comment (ayeshLK) [inline ballerina/tests/test_client.bal:462]:
```suggestion
            if acl_item.id == createdAclRule.id {

```

File: ballerina/tests/test_client.bal [hunk 16]
```diff
@@ -0,0 +1,693 @@
...
+        end: {
+            dateTime: "2023-10-28T03:30:00+05:30",
+            timeZone: "Asia/Colombo"
+        },
+        summary: eventSummary
+    };
+    Event createdEvent = check client1->/calendars/[<string>createdCal.id]/events.post(event);
+    Event moveEvent = check client1->/calendars/[<string>createdCal.id]/events/[<string>createdEvent.id]/move.post(<string>createdCal2.id);
+    test:assertEquals(moveEvent.summary, eventSummary);
+    Error? res = check client1->/calendars/[<string>createdCal2.id]/events/[<string>createdEvent.id].delete();
```
Comment (ayeshLK) [inline ballerina/tests/test_client.bal:532]: Check for other places as well.
```suggestion
    check client1->/calendars/[<string>createdCal2.id]/events/[<string>createdEvent.id].delete();

```

File: ballerina/tests/test_client.bal [hunk 17]
```diff
@@ -0,0 +1,693 @@
...
+
+@test:Config{}
+function testGetCalendarList() returns Error?  {
+    Client client1 = check new(config);
+    CalendarList calendarList = check client1->/users/me/calendarList.get();
+    test:assertNotEquals(calendarList, ());
+    CalendarListEntry[]? calendarListEntries = calendarList.items;
+    if calendarListEntries !is () {
+        test:assertTrue(calendarListEntries.length() > 0);
+    }
```
Comment (ayeshLK) [inline ballerina/tests/test_client.bal:648]:
```suggestion
    test:assertTrue(calendarListEntries is () || calendarListEntries.length() > 0);

```

Comment (Nuvindu) [reply ballerina/tests/test_client.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: ballerina/tests/test_client.bal (hunk 16 above)
Comment (Nuvindu) [reply ballerina/tests/test_client.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: ballerina/tests/test_client.bal (hunk 14 above)
Comment (Nuvindu) [reply ballerina/tests/test_client.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: ballerina/tests/test_client.bal [hunk 18]
```diff
@@ -0,0 +1,693 @@
...
+function testCreateAndDeleteCalendar() returns Error? {
+    Client client1 = check new(config);
+    string summary = "Test Calendar 1";
+    Calendar cal = {
+        summary: summary
+    };
+    Calendar unionResult = check client1->/calendars.post(cal);
+    test:assertEquals(unionResult.summary, summary);
+
+    Error? res = check client1->/calendars/[<string>unionResult.id].delete();
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply ballerina/tests/test_client.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: docs/setup/setup.md
[the remark this answers is not in the document]
Comment (Nuvindu) [reply docs/setup/setup.md:1]: Fixed in [5b49b50](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5b49b5063ecc6ef475556fe28422de7c02713723)

File: docs/oas/openapi.yaml
[the remark this answers is not in the document]
Comment (Nuvindu) [reply docs/oas/openapi.yaml:1]: Fixed in [5b49b50](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5b49b5063ecc6ef475556fe28422de7c02713723)

File: gradle.properties (hunk 13 above)
Comment (Nuvindu) [reply gradle.properties]: Fixed in [53cae63](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/53cae6393e24c09f4b87e8f27a6a6a4b2c6c4a15)

File: examples/work-schedule/main.bal (hunk 12 above)
Comment (Nuvindu) [reply examples/work-schedule/main.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: examples/project-management/main.bal [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+configurable string refreshToken = os:getEnv("REFRESH_TOKEN");
+configurable string refreshUrl = os:getEnv("REFRESH_URL");
+
+public function main() returns error? {
+    calendar:Client calendarClient = check new ({
+        auth: {
+            clientId: clientId,
+            clientSecret: clientSecret,
+            refreshToken: refreshToken,
+            refreshUrl: refreshUrl
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply examples/project-management/main.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: ballerina/utils.bal [hunk 20]
```diff
@@ -15,206 +15,200 @@
...
+#
+# + arrayName - Name of the field with arrays
+# + anyArray - Array to be serialized
+# + style - Defines how multiple values are delimited
+# + explode - Specifies whether arrays and objects should generate separate parameters
+# + return - Serialized array as a string
+isolated function getSerializedArray(string arrayName, anydata[] anyArray, string style = "form", boolean explode = true) returns string {
+    string key = arrayName;
+    string[] arrayValues = [];
+    if (anyArray.length() > 0) {
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply ballerina/utils.bal]: Fixed in [b69db94](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/b69db948da3331ce88b715df3dc2b11b780d8c57)

File: ballerina/utils.bal [hunk 21]
```diff
@@ -15,206 +15,200 @@
...
+    # Specifies whether arrays and objects should generate as separate fields
+    boolean explode = true;
+    # Specifies the custom content type
+    string contentType?;
+    # Specifies the custom headers
+    map<any> headers?;
+};
+
+enum EncodingStyle {
+    DEEPOBJECT, FORM, SPACEDELIMITED, PIPEDELIMITED
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply ballerina/utils.bal]: Fixed in [b69db94](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/b69db948da3331ce88b715df3dc2b11b780d8c57)

File: ballerina/tests/test_client.bal [hunk 22]
```diff
@@ -0,0 +1,693 @@
...
+configurable string clientSecret = os:getEnv("CLIENT_SECRET");
+configurable string refreshToken = os:getEnv("REFRESH_TOKEN");
+configurable string refreshUrl = os:getEnv("REFRESH_URL");
+
+ConnectionConfig config = {
+    auth: {
+        clientId: clientId,
+        clientSecret: clientSecret,
+        refreshToken: refreshToken,
+        refreshUrl: refreshUrl
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply ballerina/tests/test_client.bal]: Fixed in [5ec2e48](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/5ec2e48c2cbc1fa15e119f5cee8d8021de16eae5)

File: README.md [hunk 23]
```diff
@@ -1,52 +1,234 @@
...
+      // create a calendar
+      calendar:Calendar calendar = check calendarClient->/calendars.post({
+         summary: "Work Schedule"
+      });
+
+      // quick add new event
+      string eventTitle = "Sample Event";
+      calendar:Event event = check calendarClient->/calendars/[calendarId]/events/quickAdd.post(eventTitle);
+   }
+   ```
```
Comment (ThisaruGuruge) [inline README.md:132]: Check the indentations of the sample codes in MD files.

```suggestion
```ballerina
public function main() returns error? {
    calendar:Client calendarClient = ...//

    // create a calendar
    calendar:Calendar calendar = check calendarClient->/calendars.post({
        summary: "Work Schedule"
    });

    // quick add new event
    string eventTitle = "Sample Event";
    calendar:Event event = check calendarClient->/calendars/[calendarId]/events/quickAdd.post(eventTitle);
}
```
```

File: build-config/resources/Ballerina.toml [hunk 24]
```diff
@@ -0,0 +1,16 @@
+[package]
+distribution = "2201.8.2"
+org = "ballerinax"
+name = "googleapis.calendar"
+version = "@toml.version@"
+license = ["Apache-2.0"]
+authors = ["Ballerina"]
+keywords = ["Productivity/Calendars", "Cost/Free", "Vendor/Google"]
```
[the remark this answers is not in the document]
Comment (Nuvindu) [reply build-config/resources/Ballerina.toml]: Fixed in [3c9018a](https://github.com/ballerina-platform/module-ballerinax-googleapis.calendar/pull/134/commits/3c9018a6c60388808b166b86e2de1706a9281511)

**Review by ThisaruGuruge — APPROVED** [review]
